File: php/src/VcrBuilderBase.php
<?php

declare(strict_types=1);

namespace Servirtium;

use FFI\CData;

/**
 * Shared bind options for both builders.
 *
 * @template-covariant TSelf of VcrBuilderBase
 */
abstract class VcrBuilderBase
{
    protected string $tapePath;
    protected string $hostValue = '127.0.0.1';
    protected int $portValue = 0;          // 0 => OS-assigned (dynamic)
    protected string $labelValue = '';

    /** @var list<array{0: VcrField, 1: string}> */
    private array $headerRemovals = [];

    /** @var list<array{0: string, 1: string}> */
    private array $staticContent = [];

    /** @var list<string> */
    private array $untaped = [];

    protected function __construct(string $tapePath)
    {
        $this->tapePath = $tapePath;
    }

    /** Bind host. Defaults to 127.0.0.1. */
    public function host(string $host): static
    {
        $this->hostValue = $host;

        return $this;
    }

    /** Bind port. 0 (the default) asks the OS for a free port. */
    public function port(int $port): static
    {
        $this->portValue = $port;

        return $this;
    }

    /** Human-facing label for logs/diagnostics (not a state key). */
    public function label(string $label): static
    {
        $this->labelValue = $label;

        return $this;
    }

    /** Remove a header by name from the given block (case-insensitive). */
    public function removeHeader(VcrField $field, string $name): static
    {
        $this->headerRemovals[] = [$field, $name];

        return $this;
    }

    /**
     * Serve a path prefix from an on-disk directory instead of the tape
     * (Servirtium step 11).
     *
     * Works in both playback and record mode — recording a browser suite is
     * cleaner served same-origin from the VCR (no CORS preflights).
     */
    public function staticContent(string $mountPath, string $fsDir): static
    {
        $this->staticContent[] = [$mountPath, $fsDir];

        return $this;
    }

    /**
     * Mark an incidental path (e.g. /favicon.ico) the VCR answers 404 for
     * without touching the tape — no cursor consumed on playback, nothing
     * forwarded or recorded on record.
     */
    public function untaped(string $path): static
    {
        $this->untaped[] = $path;

        return $this;
    }

    /**
     * Apply this builder's accumulated config to the opened handle, before
     * serving starts. Subclasses extend this.
     */
    protected function applyConfig(CData $handle): void
    {
        $lib = Native::lib();
        foreach ($this->headerRemovals as [$field, $name]) {
            self::check($lib->aether_vcr_embed_remove_header($handle, $field->value, $name), 'removeHeader');
        }
        foreach ($this->staticContent as [$mount, $dir]) {
            self::check($lib->aether_vcr_embed_static_content($handle, $mount, $dir), 'staticContent');
        }
        foreach ($this->untaped as $path) {
            self::check($lib->aether_vcr_embed_untaped($handle, $path), 'untaped');
        }
    }

    /** Throw if a mutation call returned a non-empty error string ("" = success). */
    final protected static function check(?CData $resultPtr, string $op): void
    {
        $err = Native::takeString($resultPtr);
        if ($err !== '') {
            throw new VcrException("vcr {$op} failed: {$err}");
        }
    }

    /** Drain the last-error slot for a failed start, with a fallback hint. */
    final protected static function drainStartError(CData $handle): string
    {
        $err = Native::takeString(Native::lib()->aether_vcr_embed_last_error($handle));

        return $err !== '' ? $err : '(no detail; check tape path and port availability)';
    }
}

File: java/src/main/java/com/paulhammant/servirtium/vcr/VcrBuilderBase.java
package com.paulhammant.servirtium.vcr;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.util.ArrayList;
import java.util.List;

/**
 * Shared bind options and per-handle config handling for both builders.
 *
 * @param <S> the concrete builder type, for fluent self-returns.
 */
public abstract class VcrBuilderBase<S extends VcrBuilderBase<S>> {

    record HeaderRemoval(Field field, String name) {
    }

    private record StaticMount(String mount, String dir) {
    }

    final String tapePath;
    String host = "127.0.0.1";
    int port;            // 0 => OS-assigned (dynamic)
    String label = "";

    private final List<HeaderRemoval> headerRemovals = new ArrayList<>();
    private final List<StaticMount> staticContent = new ArrayList<>();
    private final List<String> untaped = new ArrayList<>();

    VcrBuilderBase(String tapePath) {
        this.tapePath = tapePath;
    }

    abstract S self();

    /** Bind host. Defaults to 127.0.0.1. */
    public S host(String host) {
        this.host = host;
        return self();
    }

    /** Bind port. 0 (the default) asks the OS for a free port. */
    public S port(int port) {
        this.port = port;
        return self();
    }

    /** Human-facing label for logs/diagnostics (not a state key). */
    public S label(String label) {
        this.label = label;
        return self();
    }

    /** Remove a header by name from the given block (case-insensitive). */
    public S removeHeader(Field field, String name) {
        headerRemovals.add(new HeaderRemoval(field, name));
        return self();
    }

    /**
     * Serve a path prefix from an on-disk directory instead of the tape
     * (Servirtium step 11).
     *
     * <p>Works in both playback and record mode (recording a browser suite is
     * cleaner served same-origin from the VCR — no CORS preflights).
     */
    public S staticContent(String mountPath, String fsDir) {
        staticContent.add(new StaticMount(mountPath, fsDir));
        return self();
    }

    /**
     * Mark an incidental request path (e.g. {@code /favicon.ico}) so the VCR
     * answers 404 without consuming the tape cursor (playback) and never
     * forwards/records it (record).
     */
    public S untaped(String path) {
        untaped.add(path);
        return self();
    }

    /**
     * Apply this builder's accumulated config to the opened handle, before
     * serving starts. Subclasses extend this.
     */
    void applyConfig(Arena arena, MemorySegment handle) {
        for (HeaderRemoval h : headerRemovals) {
            try {
                MemorySegment r = (MemorySegment) NativeMethods.REMOVE_HEADER.invokeExact(
                        handle, h.field().code(), NativeMethods.cString(arena, h.name()));
                check(r, "removeHeader");
            } catch (Throwable t) {
                throw rethrow("removeHeader", t);
            }
        }
        for (StaticMount s : staticContent) {
            try {
                MemorySegment r = (MemorySegment) NativeMethods.STATIC_CONTENT.invokeExact(
                        handle,
                        NativeMethods.cString(arena, s.mount()),
                        NativeMethods.cString(arena, s.dir()));
                check(r, "staticContent");
            } catch (Throwable t) {
                throw rethrow("staticContent", t);
            }
        }
        for (String p : untaped) {
            try {
                MemorySegment r = (MemorySegment) NativeMethods.UNTAPED.invokeExact(
                        handle, NativeMethods.cString(arena, p));
                check(r, "untaped");
            } catch (Throwable t) {
                throw rethrow("untaped", t);
            }
        }
    }

    /** Throw if a mutation call returned a non-empty error string ("" = success). */
    static void check(MemorySegment resultPtr, String op) {
        String err = NativeMethods.takeString(resultPtr);
        if (err != null && !err.isEmpty()) {
            throw new VcrException("vcr " + op + " failed: " + err);
        }
    }

    static VcrException rethrow(String op, Throwable t) {
        if (t instanceof VcrException ve) {
            return ve;
        }
        return new VcrException("vcr " + op + " failed: " + t.getMessage());
    }

    static String drainStartError(MemorySegment handle) {
        try {
            MemorySegment p = (MemorySegment) NativeMethods.LAST_ERROR.invokeExact(handle);
            String err = NativeMethods.takeString(p);
            return (err == null || err.isEmpty())
                    ? "(no detail; check tape path and port availability)"
                    : err;
        } catch (Throwable t) {
            return "(could not read last_error: " + t.getMessage() + ")";
        }
    }
}
